lib/Base/NumpyReader.cpp

```diff
@@ -0,0 +1,324 @@
+    break;
+  case Int32Ty:
+    buffertoTypedVector<int32_t>(elem_size, nvals, databuf, data);
+    break;
+  case UInt32Ty:
+    buffertoTypedVector<uint32_t>(elem_size, nvals, databuf, data);
+    break;
+  case Int64Ty:
+    buffertoTypedVector<int64_t>(elem_size, nvals, databuf, data);
+    break;
+  case UInt64Ty:
+    buffertoTypedVector<uint64_t>(elem_size, nvals, databuf, data);
+    break;
+  case Float32Ty:
+    buffertoTypedVector<float>(elem_size, nvals, databuf, data);
+    break;
+  case Float64Ty:
+    buffertoTypedVector<double>(elem_size, nvals, databuf, data);
+    break;
+  default:
+    LOG(FATAL) << "NPY loader: Not supported type: " << type;
+  }
+  delete[] buffer;
+}
+
+void glow::loadNumpyImagesAndPreprocess(
+    const llvm::ArrayRef<std::string> &filenames, Tensor &inputData,
+    ImageNormalizationMode imageNormMode, ImageLayout imageLayout,
+    ImageLayout inputLayout, llvm::ArrayRef<float> mean,
+    llvm::ArrayRef<float> stddev) {
+  DCHECK(!filenames.empty())
+      << "NPY loader: There must be at least one filename in filenames.";
+  dim_t numImg = filenames.size();
+
+  auto normalizeData = [&mean, &stddev, &imageLayout,
+                        &imageNormMode](std::vector<float> &data,
+                                        std::vector<dim_t> &shape) {
+    auto range = normModeToRange(imageNormMode);
+    float scale = ((range.second - range.first) / 255.f);
+    float bias = range.first;
+    dim_t numCh = (imageLayout == ImageLayout::NHWC) ? shape[3] : shape[1];
+    std::vector<float> zeroMean(numCh, 0.f);
+    std::vector<float> oneStd(numCh, 1.f);
+    std::vector<float> meanVal(mean.size() ? mean
+                                           : llvm::makeArrayRef(zeroMean));
+    std::vector<float> stddevVal(stddev.size() ? stddev
+                                               : llvm::makeArrayRef(oneStd));
+    CHECK_EQ(numCh, meanVal.size()) << "NPY loader: mean argument size should "
+                                       "match the number of channels: "
+                                    << numCh;
+    CHECK_EQ(numCh, stddevVal.size()) << "NPY loader: stddev argument size "
+                                         "should match the number of channels: "
+                                      << numCh;
+    size_t chStride =
+        (imageLayout == ImageLayout::NHWC) ? 1 : shape[2] * shape[3];
+    for (size_t i = 0; i < data.size(); i++) {
+      size_t chIdx = (i / chStride) % numCh;
+      CHECK(data[i] >= 0. && data[i] <= 255.)
+          << "NPY loader: U8 data expected, got: " << data[i];
+      data[i] = (data[i] - meanVal[chIdx]) / stddevVal[chIdx];
+      data[i] = data[i] * scale + bias;
+    }
+  };
+
+  // Read each tensor file into a vector of tensors.
+  std::vector<Tensor> tensors(numImg);
+  dim_t batchSize = 0;
+  for (dim_t n = 0; n < numImg; n++) {
+    std::vector<dim_t> shape;
+    std::vector<float> data;
+    numpyReader(filenames[n], shape, data);
```

Comment:
So we always use data with float -- do you have private uses for other datatypes? Wondering because `buffertoTypedVector()` has support for other types but doesn't appear we're using/testing that anywhere.